aptos-move/move-examples/mint_nft/sources/minting.move

```diff
@@ -0,0 +1,277 @@
+        let resource_signer = account::create_signer_with_capability(&resource_signer_cap);
+
+        // create the nft collection
+        let maximum_supply = 0;
+        let mutate_setting = vector<bool>[ false, false, false ];
+        let resource_account_address = signer::address_of(&resource_signer);
+        token::create_collection(&resource_signer, collection_name, description, collection_uri, maximum_supply, mutate_setting);
+
+        // create a token data id to specify which token will be minted
+        let token_data_id = token::create_tokendata(
+            &resource_signer,
+            collection_name,
+            token_name,
+            string::utf8(b""),
+            0,
+            token_uri,
+            resource_account_address,
+            0,
+            0,
+            // we don't allow any mutation to the token
+            token::create_token_mutability_config(
+                &vector<bool>[ false, false, false, false, true ]
+            ),
+            vector::empty<String>(),
+            vector::empty<vector<u8>>(),
+            vector::empty<String>(),
+        );
+
+        let public_key = std::option::extract(&mut ed25519::new_validated_public_key_from_bytes(public_key_bytes));
+
+        move_to(resource_account, CollectionTokenMinter {
+            public_key,
+            signer_cap: resource_signer_cap,
+            token_data_id,
+            expiration_timestamp,
+            minting_enabled: true,
+            token_minting_events: account::new_event_handle<TokenMintingEvent>(&resource_signer),
+        });
+    }
+
+    /// Set if minting is enabled for this collection token minter
+    public entry fun set_minting_enabled(minter: &signer, minting_enabled: bool) acquires CollectionTokenMinter {
+        let minter_address = signer::address_of(minter);
+        assert!(minter_address == @mint_nft, error::permission_denied(ENOT_AUTHORIZED));
+        let collection_token_minter = borrow_global_mut<CollectionTokenMinter>(minter_address);
+        collection_token_minter.minting_enabled = minting_enabled;
+    }
+
+    /// Set the expiration timestamp of this collection token minter
+    public entry fun set_timestamp(minter: &signer, expiration_timestamp: u64) acquires CollectionTokenMinter {
+        let minter_address = signer::address_of(minter);
+        assert!(minter_address == @mint_nft, error::permission_denied(ENOT_AUTHORIZED));
+        let collection_token_minter = borrow_global_mut<CollectionTokenMinter>(minter_address);
+        collection_token_minter.expiration_timestamp = expiration_timestamp;
+    }
+
+    /// Mint an NFT to the receiver.
+    /// `mint_proof_signature` should be the `MintProofChallenge` signed by the resource signer's private key
+    /// `public_key_bytes` should be the public key of the resource signer
+    public entry fun mint_nft(receiver: &signer, mint_proof_signature: vector<u8>) acquires CollectionTokenMinter {
```

Comment:
The only issue I have with this API is that by not passing in the actual message, we cannot validate why it failed. Granted... we don't have bcs bytes to struct...

this is good for now, we need to improve that aspect.

then we could give users better errors on why their txn failed

Reply:
ahh yeah good point!